projects/3-mazes/neil.ipynb

```diff
@@ -0,0 +1,808 @@
+    "\n",
+    "Try to get this assignment done by Friday; the other two may take a little longer but the sooner you attempt them the earlier you can get feedback!\n",
+    "\n",
+    "You may also try visualizing paths through the maze, implementing several different heuristic searches, comparing against aheuristic search, etc.\n",
+    "\n",
+    "Generating mazes automatically would also be a great exercise!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 52,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "def manhattan_distance(point_a, point_b):\n",
+    "    (x0,y0) = point_a\n",
+    "    (x1,y1) = point_b\n",
+    "    return abs(y1-y0) + abs(x1-x0)    \n",
+    "\n",
+    "def element_wise_tuple_addition(x, y):\n",
+    "    z = []\n",
+    "    for i in range(len(x)):\n",
+    "        z.append(x[i] + y[i])\n",
+    "    return tuple(z)\n",
+    "\n",
+    "\n",
+    "def heuristic(maze):\n",
+    "    frontier = Q.PriorityQueue()\n",
+    "    \n",
+    "    frontier.put((0, maze))\n",
+    "    #came_from = {}\n",
+    "    #cost_so_far = {}\n",
+    "    \n",
+    "    #came_from[start] = None\n",
+    "    #cost_so_far[start] = 0\n",
+    "    visited = set([maze])\n",
+    "    \n",
+    "    while not frontier.empty():\n",
+    "        (priority, current) = frontier.get() # Maze\n",
+    "        \n",
+    "        #print(current)\n",
+    "        \n",
+    "        #print(current.available_moves())\n",
+    "        if current.is_at_exit():\n",
+    "            print(\"Returning path because at \" + str(current.player_pos) + \" while goal is \" + str(current.exit_pos))\n",
+    "            current.path.pop()\n",
+    "            return current.path\n",
+    "\n",
+    "        for move in current.available_moves():\n",
+    "            #print(\"Checking move \" + str(move))\n",
+    "            \n",
+    "            successor = current.clone()\n",
+    "            \n",
+    "            if move == \"switch\":\n",
+    "                #print(\"Switching at \" + str(successor.player_pos))\n",
+    "                successor.toggle()\n",
+    "                \n",
+    "         \n",
+    "                \n",
+    "                #print(\"Successfully toggled\")\n",
+    "            else:\n",
+    "                (x,y) = move\n",
+    "                successor.move_player(x,y)\n",
+    "            \n",
+    "            #if not current in cost_so_far:\n",
+    "                #print(\"Not found, defaulting to 1\")\n",
+    "                #cost_so_far[current] = 1\n",
+    "            \n",
+    "            #new_cost = cost_so_far[current] + 1\n",
+    "            new_cost = current.cost + 1\n",
+    "            \n",
+    "            if not successor in visited or new_cost < current.cost:\n",
```

Comment:
This `or` isn't quite right, and turns this into `if not in visited or false` since `(current.cost+1)` is never less than `current.cost`. You want to compare against the cost of the member of visited which is equivalent to `successor`. This could lead to finding suboptimal paths since the node equivalent to `successor` may be tagged with a higher cost but never replaced.